internal/scriptlet/addrule.go

```diff
@@ -0,0 +1,147 @@
+package scriptlet
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"regexp"
+	"strings"
+)
+
+var (
+	// reAdguardScriptlet detects and extracts key data from AdGuard-style scriptlets.
+	reAdguardScriptlet = regexp.MustCompile(`(.*)#%#\/\/scriptlet\((.+)\)`)
+	// adguardToCanonical maps AdGuard scriptlet names to their respective implementations inside the scriptlet bundle.
+	adguardToCanonical = map[string]string{
+		"set-local-storage-item":   "setLocalStorageItem",
+		"set-session-storage-item": "setSessionStorageItem",
+		"nowebrtc":                 "nowebrtc",
+		"prevent-fetch":            "preventFetch",
+		"prevent-xhr":              "preventXHR",
+		"set-constant":             "setConstant",
+	}
+	// reUblockScriptlet detects and extracts key data from uBlock Origin-style scriptlets.
+	reUblockScriptlet = regexp.MustCompile(`(.*)##\+js\((.+)\)`)
+	// ublockToCanonical maps uBlock Origin scriptlet names to their respective implementations inside the scriptlet bundle.
+	ublockToCanonical = map[string]string{
+		// TODO: manually check ublock syntax compatibility
+		"set-local-storage-item": "setLocalStorageItem",
+		"no-xhr-if":              "preventXHR",
+		"no-fetch-if":            "preventFetch",
+		"nowebrtc":               "nowebrtc",
+		"set-constant":           "setConstant",
+	}
+	errNotQuotedString    = errors.New("not a quoted string")
+	errUnsupportedSyntax  = errors.New("unsupported syntax")
+	errEmptyScriptletBody = errors.New("scriptlet body is empty")
+)
+
+func (i *Injector) AddRule(rule string) error {
+	var rawHostnames string
+	var scriptlet *Scriptlet
+	var err error
+	if match := reAdguardScriptlet.FindStringSubmatch(rule); match != nil {
+		rawHostnames = match[1]
+		scriptlet, err = parseAdguardScriptlet(match[2])
+		if err != nil {
+			return fmt.Errorf("parse adguard scriptlet: %w", err)
+		}
+	} else if match := reUblockScriptlet.FindStringSubmatch(rule); match != nil {
+		rawHostnames = match[1]
+		scriptlet, err = parseUblockScriptlet(match[2])
+		if err != nil {
+			return fmt.Errorf("parse ublock origin scriptlet: %w", err)
+		}
+	} else {
+		return errUnsupportedSyntax
+	}
+
+	if len(rawHostnames) == 0 {
```

Comment:
Could you please explain the reason for adding a scriptlet into the store without a hostname?

Reply:
Scriptlets without hostnames are valid, and should be applied at any webpage: https://adguard.com/kb/general/ad-filtering/create-own-filters/#scriptlets (under "Blocking rules syntax").

Reply:
Thank you for clarifiction.